Add facings support and other controls to LeavesTrails

reaperrr opened the pull request:
Closes #3188.

reaperrr committed:
Add facings support to SpriteEffect

reaperrr committed:
Add facings support, start delay and offset controls to LeavesTrails

reaperrr committed:
Add control over whether trails should spawn with or without delay

pchote commented on OpenRA.Mods.Common/Traits/Render/LeavesTrails.cs:
```diff
@@ -49,41 +49,89 @@ public class LeavesTrailsInfo : UpgradableTraitInfo
+				cachedFacing = facing != null ? facing.Facing : 0;
 				ticks = 0;
+
+				if (!evenNumber)
```
evenNumber ^= true;

pchote commented on OpenRA.Mods.Common/Traits/Render/LeavesTrails.cs:
```diff
@@ -49,41 +49,89 @@ public class LeavesTrailsInfo : UpgradableTraitInfo
+		public readonly WVec Offset = WVec.Zero;
+
+		[Desc("Use opposite offset for every second spawned trail.")]
+		public readonly bool AlternateOffset = false;
```
A list of offsets that are looped might be cleaner and more general.

reaperrr committed:
Make LeavesTrails loop through a list of offsets

Instead of the unflexible 'AlternateOffset' approach.

reaperrr commented:
I get an IndexOutOfRangeException with my attempt at looping through offsets, not sure what I need to change to fix that.

pchote commented on OpenRA.Mods.Common/Traits/Render/LeavesTrails.cs:
```diff
@@ -109,11 +106,15 @@ public void Tick(Actor self)
 
-				var offsetRotation = Info.Offset.Rotate(body.QuantizeOrientation(self, self.Orientation));
-				var spawnOffset = body.LocalToWorld(Info.AlternateOffset && evenNumber ? -offsetRotation : offsetRotation);
+				if (offset < Info.Offsets.Length)
```
If `offset == Info.Offsets.Length - 1` it will pass this check and be incremented to `Info.Offsets.Length`, which then gives your OutOfBoundsException.

You can simplify this if/else block to:

```csharp
if (++offset >= Info.Offsets.Length)
    offset = 0;
```

pchote commented:
Looks like Github is somehow screwing up commit-level comments, thinking that they're outdated.  See my comment just above for a fix.

reaperrr commented:
@pchote I'm still getting
```
Exception of type `System.IndexOutOfRangeException`: Der Index war außerhalb des Arraybereichs.
   bei OpenRA.Mods.Common.Traits.LeavesTrails.Tick(Actor self)
   bei OpenRA.World.<Tick>b__7(TraitPair`1 x) in e:\ORA_work_coding\OpenRA.Game\World.cs:Zeile 329.
   bei OpenRA.WorldUtils.DoTimed[T](IEnumerable`1 e, Action`1 a, String text) in e:\ORA_work_coding\OpenRA.Game\WorldUtils.cs:Zeile 70.
   bei OpenRA.World.Tick() in e:\ORA_work_coding\OpenRA.Game\World.cs:Zeile 0.
   bei OpenRA.Game.InnerLogicTick(OrderManager orderManager) in e:\ORA_work_coding\OpenRA.Game\Game.cs:Zeile 522.
   bei OpenRA.Game.LogicTick() in e:\ORA_work_coding\OpenRA.Game\Game.cs:Zeile 547.
   bei OpenRA.Game.Loop() in e:\ORA_work_coding\OpenRA.Game\Game.cs:Zeile 674.
   bei OpenRA.Game.Run() in e:\ORA_work_coding\OpenRA.Game\Game.cs:Zeile 714.
   bei OpenRA.Program.Run(String[] args) in e:\ORA_work_coding\OpenRA.Game\Support\Program.cs:Zeile 118.
   bei OpenRA.Program.Main(String[] args) in e:\ORA_work_coding\OpenRA.Game\Support\Program.cs:Zeile 40.
```

with your fix.

reaperrr commented:
Nevermind, silly me had a typo in the test case and no default value in `Offsets`.

reaperrr commented:
Updated, works now.

Mailaender commented:
Confirmed, works as promised. :white_check_mark:

reaperrr commented:
On a side-note, the facings support added to `SpriteEffect` will allow us to get rid of both `Smoke` and `Explosion` effects since they feature nothing exclusive and are effectively just more limited reverse-duplicates of `SpriteEffect`.
That would also have the positive effect of a consistent parameter order, which is currently inconsistent between all 3, and allow directional explosion effects down the line.

abcdefg30 commented:
Needs a rebase.

reaperrr commented:
Rebased. Since the test case was responsible for the need to rebase, I moved it to a separate branch:
https://github.com/reaperrr/OpenRA/tree/improve-LeavesTrails-test

obrakmann commented on OpenRA.Game/Effects/SpriteEffect.cs:
```diff
@@ -23,14 +23,14 @@ public class SpriteEffect : IEffect
 		readonly bool scaleSizeWithZoom;
 
-		public SpriteEffect(WPos pos, World world, string image, string sequence, string palette, bool visibleThroughFog = false, bool scaleSizeWithZoom = false)
+		public SpriteEffect(WPos pos, World world, string image, string sequence, string palette, int facing = 0, bool visibleThroughFog = false, bool scaleSizeWithZoom = false)
```
Is there a special reason that you're adding the `facing` parameter smack in the middle? If you'd put it on the end, people (well, those few with 3rd party dlls) wouldn't need to adjust their code.

reaperrr replied:
Well, one thing I had in mind was to later ditch inferior special case effects like `Explosion`, `Smoke` and so on and use `SpriteEffect` for those as well, and I wanted to avoid having to explicitly set `false` for the other two optional parameters. But I guess you're right, downstream compatibility is a bit more important.

obrakmann commented:
Looks ok otherwise I'm really looking forward to the eye candy this will bring. :+1:

reaperrr committed:
Move SpriteEffect facing parameter to the end

To avoid potential incompatibilities with downstream projects.

reaperrr commented:
Updated.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/4da65db625de26ba0629bab8f8166c9dcb989e5b)
